platform/cc/RuntimeEffect.cc

```diff
@@ -0,0 +1,43 @@
+#include <jni.h>
+
+#include "SkRuntimeEffect.h"
+#include "interop.hh"
+
+JNIEXPORT jlong JNICALL
+Java_org_jetbrains_skija_RuntimeEffect__1nMakeShader(JNIEnv* env,
+                                                     jclass jclass,
+                                                     jlong ptr,
+                                                     jlong uniformPtr,
+                                                     jlongArray childrenPtrs,
+                                                     jlong childCount,
+                                                     jfloatArray localMatrixArr,
+                                                     jboolean isOpaque) {
+    SkRuntimeEffect* runtimeEffect = jlongToPtr<SkRuntimeEffect*>(ptr);
+    SkData* uniform = jlongToPtr<SkData*>(uniformPtr);
+    sk_sp<SkData> uniformData = SkData::MakeFromMalloc(uniform, uniform->size());
+
+    // Matrix
+    jfloat* m = env->GetFloatArrayElements(localMatrixArr, 0);
+    SkMatrix* mPtr = new SkMatrix();
+    mPtr->setAll(m[0], m[1], m[2], m[3], m[4], m[5], m[6], m[7], m[8]);
+    env->ReleaseFloatArrayElements(localMatrixArr, m, 0);
+
+    sk_sp<SkShader>* children = new sk_sp<SkShader>[childCount];
+    SkShader** cPtrs = reinterpret_cast<SkShader**>(childrenPtrs);
```

Comment:
This won’t fly. Java arrays are not C arrays, unfortunately. You need to request special access to them. You need env->GetLongArrayElements, paired with ReleaseLongArrayElements after the loop. Each individual long can then be cast to pointer using jlongToPtr

Reply:
Here’s an example of a similar situation https://github.com/JetBrains/skija/blob/8b2111460148cf6f190e5a1db24cf8ebfb069dcd/platform/cc/ImageFilter.cc#L126-L139